v2/routing.go

```diff
@@ -110,20 +111,100 @@ func (d *DHT) Provide(ctx context.Context, c cid.Cid, brdcst bool) error {
 }
 
 func (d *DHT) FindProvidersAsync(ctx context.Context, c cid.Cid, count int) <-chan peer.AddrInfo {
-	_, span := d.tele.Tracer.Start(ctx, "DHT.FindProvidersAsync", otel.WithAttributes(attribute.String("cid", c.String()), attribute.Int("count", count)))
+	peerOut := make(chan peer.AddrInfo)
+	go d.findProvidersAsyncRoutine(ctx, c, count, peerOut)
+	return peerOut
+}
+
+func (d *DHT) findProvidersAsyncRoutine(ctx context.Context, c cid.Cid, count int, out chan peer.AddrInfo) {
+	_, span := d.tele.Tracer.Start(ctx, "DHT.findProvidersAsyncRoutine", otel.WithAttributes(attribute.String("cid", c.String()), attribute.Int("count", count)))
 	defer span.End()
 
-	// verify if this DHT supports provider records by checking if a "providers"
-	// backend is registered.
-	_, found := d.backends[namespaceProviders]
+	defer close(out)
+
+	// verify if this DHT supports provider records by checking
+	// if a "providers" backend is registered.
+	b, found := d.backends[namespaceProviders]
 	if !found || !c.Defined() {
-		peerOut := make(chan peer.AddrInfo)
-		close(peerOut)
-		return peerOut
+		span.RecordError(fmt.Errorf("no providers backend registered or CID undefined"))
+		return
 	}
 
-	// TODO reach out to Zikade
-	panic("implement me")
+	// first fetch the record locally
+	stored, err := b.Fetch(ctx, string(c.Hash()))
+	if err != nil {
```

Comment:
This would be necessary if this was the datastore Get method. This is the backend Fetch method. The backend returns a `*providerSet`. If there are no providers, the set will be empty. IIRC internally the backend will _query_ the datastore (not `Get`). This means even the datastore won't return a NotFound error.